src/PrestaShopBundle/Routing/ControllerStack.php

```diff
@@ -0,0 +1,102 @@
+<?php
+/**
+ * 2007-2020 PrestaShop SA and Contributors
+ *
+ * NOTICE OF LICENSE
+ *
+ * This source file is subject to the Open Software License (OSL 3.0)
+ * that is bundled with this package in the file LICENSE.txt.
+ * It is also available through the world-wide-web at this URL:
+ * https://opensource.org/licenses/OSL-3.0
+ * If you did not receive a copy of the license and are unable to
+ * obtain it through the world-wide-web, please send an email
+ * to license@prestashop.com so we can send you a copy immediately.
+ *
+ * DISCLAIMER
+ *
+ * Do not edit or add to this file if you wish to upgrade PrestaShop to newer
+ * versions in the future. If you wish to customize PrestaShop for your
+ * needs please refer to https://www.prestashop.com for more information.
+ *
+ * @author    PrestaShop SA <contact@prestashop.com>
+ * @copyright 2007-2020 PrestaShop SA and Contributors
+ * @license   https://opensource.org/licenses/OSL-3.0 Open Software License (OSL 3.0)
+ * International Registered Trademark & Property of PrestaShop SA
+ */
+
+declare(strict_types=1);
+
+namespace PrestaShopBundle\Routing;
+
+use Symfony\Bundle\FrameworkBundle\Controller\Controller;
+
+/**
+ * Inspired by \Symfony\Component\HttpFoundation\RequestStack
+ */
+class ControllerStack
+{
+    /**
+     * @var Controller[]
+     */
+    private $controllers = [];
```

Comment:
The class is not final, don't put this accessor private, otherwise, someone creating his stack must recreate properties.
```suggestion
    protected $controllers = [];
```

Reply:
🤔 Do you mean we should always introduce `protected` properties for non final classes, never `private` unless wanting to ?

Reply:
Yes, otherwise you must rewrite everything due to lack of visibility :/

Reply:
I'm actually OK with that 😄 but I'm afraid The Big Boss will be mad against me because [he said he disliked making private methods protected](https://github.com/PrestaShop/PrestaShop/issues/18901#issuecomment-623335766).

I feel like we have 2 schools on this debate: "use protected, be extensible" to allow better customization versus "use final and private, promote composition over inheritance".

@eternoendless Could you please tell us your vision on this topic ?

Reply:
By the open/closed principle, classes should be open to extension and closed to modification.
If you make a property protected, it becomes part of your contract. Are you willing to make that commitment?

Reply:
If you use private properties, you must provide accessors in order to be able to use others methods like before.
In this case, if `$controllers` is private, add `getControllers()` and use `$this->getControllers()` instead of `$this->controllers`.
Otherwise, you can extend the class, but you're not able to change anything because you can't call `$this->controllers`

Reply:
@PierreRambaud It makes sense. Nice feedback, I will apply it !

Reply:
=> next part of this discussion https://github.com/PrestaShop/PrestaShop/issues/21383